contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -0,0 +1,147 @@
+    };
+
+    let payload_hash = Keccak256::digest(payload.as_slice()).into();
+
+    let msg = Message {
+        cc_id: cc_id.clone(),
+        source_address: Address::try_from(sender.clone().into_string())
+            .expect("failed to convert sender address"),
+        destination_chain,
+        destination_address,
+        payload_hash,
+    };
+
+    state::save_incoming_msg(store, cc_id, &msg).change_context(Error::InvalidStoreAccess)?;
+
+    let (wasm_msg, events) = route(router, vec![msg.clone()])?;
+
+    Ok(Response::new()
+        .add_message(wasm_msg)
+        .add_event(AxelarnetGatewayEvent::ContractCalled { msg, payload }.into())
+        .add_events(events))
+}
+
+// Because the messages came from the router, we can assume they are already verified
+pub(crate) fn receive_messages(
+    store: &mut dyn Storage,
+    msgs: Vec<Message>,
+) -> Result<Response, Error> {
+    for msg in msgs.iter() {
+        state::save_outgoing_msg(store, msg.cc_id.clone(), msg.clone())
+            .change_context(Error::SaveOutgoingMessage)?;
+    }
+
+    Ok(Response::new().add_events(
+        msgs.into_iter()
+            .map(|msg| AxelarnetGatewayEvent::Routing { msg }.into()),
+    ))
+}
+
+pub(crate) fn send_messages(
+    store: &mut dyn Storage,
+    router: &Router,
+    msgs: Vec<Message>,
+) -> Result<Response, Error> {
+    for msg in msgs.iter() {
+        let stored_msg = state::may_load_incoming_msg(store, &msg.cc_id)
+            .change_context(Error::InvalidStoreAccess)?;
+
+        match stored_msg {
+            Some(message) if msg != &message => {
+                Err(report!(Error::MessageMismatch(msg.cc_id.clone())))
+            }
+            Some(_) => Ok(()),
+            None => Err(report!(Error::MessageNotFound(msg.cc_id.clone()))),
+        }?
+    }
+
+    let (wasm_msg, events) = route(router, msgs)?;
+
+    Ok(Response::new().add_message(wasm_msg).add_events(events))
+}
+
+fn route(
+    router: &Router,
+    msgs: Vec<Message>,
+) -> Result<(WasmMsg, impl IntoIterator<Item = Event>), Error> {
+    Ok((
+        router.route(msgs.clone()).ok_or(Error::RoutingFailed)?,
+        msgs.into_iter()
+            .map(|msg| AxelarnetGatewayEvent::Routing { msg }.into()),
+    ))
+}
+
+pub(crate) fn execute(
+    store: &mut dyn Storage,
+    api: &dyn Api,
+    querier: QuerierWrapper,
+    chain_name: ChainName,
+    cc_id: CrossChainId,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    let msg = state::update_msg_status(store, cc_id.clone())
+        .change_context(Error::MessageStatusUpdateFailed(cc_id))?;
+
+    let payload_hash: [u8; 32] = Keccak256::digest(payload.as_slice()).into();
+    if payload_hash != msg.payload_hash {
+        return Err(report!(Error::PayloadHashMismatch));
+    }
+
+    if chain_name != msg.destination_chain {
+        return Err(report!(Error::InvalidDestinationChain(
+            msg.destination_chain
+        )));
+    }
```

Comment:
Should this be an assert? How could this happen?

Reply:
Or maybe this check should be moved to `receive_messages`?

Reply:
Config issue potentially. Gateway is registered under a different chain name in the router. Since the gateway set the chain_name for send_messages, I think it's a fine sanity check to apply for receiving messages (still prefer returning a structured error vs an assert, but not a strong opinion). I prefer checking it here instead of the state file, since it's being checked against `chain_name` which `receive_messages` doesn't have internally (or it has to load the config again).

Reply:
I meant we could check this when receiving messages from the router, instead of when trying to execute. I prefer failing earlier rather than later.

Reply:
Sure, since it's a sanity check, fine to fail there